### PR TITLE
Paligemma- fix devices and dtype assignments

### LINKED ISSUE #30997: loss calculation for PaliGemmaForConditionalGeneration potentially not cast to correct device
### System Info

```
Name: transformers
Version: 4.42.0.dev0
```

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Use `PaliGemmaForConditionalGeneration.forward` with device_map='auto' and labels for 4+ devices.

### Expected behavior

I believe which device logits/labels/attention_mask will be on for `PaliGemmaForConditionalGeneration` can potentially cause an error for others (and is for me as well by default as well, but I have patched it) when using `device_map='auto'`.

https://github.com/huggingface/transformers/blob/main/src/transformers/models/paligemma/modeling_paligemma.py#L487

```
                shift_logits = shift_logits[shift_attention_mask.to(logits.device) != 0].contiguous()
                shift_labels = shift_labels[shift_attention_mask.to(logits.device) != 0].contiguous()
```

labels and attention_mask should probably be on the lowest index device.  logits likely to be a later device if the model is sharded with something like `device_map='auto'`.  The shift_logits line is right since the shift_logits come from the logits but the shift_labels line can cause issues since shift_labels should likely be moved to logits device.

### PR DESCRIPTION
**Pull request by molbap**
# What does this PR do?
Moves tensors to correct devices in case of multi-gpu training on accelerate and device_map = auto. 
Additionally ensures bf16 training works as well.

Fixes #30997

### CODE REVIEW

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 1]
```diff
@@ -484,7 +487,7 @@ def forward(
                 # we use the input attention mask to shift the logits and labels, because it is 2D.
                 shift_attention_mask = input_attention_mask[..., 1:]
                 shift_logits = shift_logits[shift_attention_mask.to(logits.device) != 0].contiguous()
-                shift_labels = shift_labels[shift_attention_mask.to(logits.device) != 0].contiguous()
+                shift_labels = shift_labels[shift_attention_mask.to(shift_labels.device) != 0].contiguous()
```
[the remark this answers is not in the document]
Comment (pcuenca) [reply src/transformers/models/paligemma/modeling_paligemma.py:490]: actually no, deleting

**Review by pcuenca — APPROVED** [review]
LGTM, thanks for the quick fix!

**Comment by molbap** [comment]
cc @ArthurZucker wdyt?

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 2]
```diff
@@ -301,14 +301,15 @@ def _merge_input_ids_with_image_features(
         pad_mask = input_ids == self.pad_token_id
 
         # expand masks to match embedding dimension
-        text_mask_expanded = text_mask.unsqueeze(-1).expand(-1, -1, embed_dim)
-        pad_mask_expanded = pad_mask.unsqueeze(-1).expand(-1, -1, embed_dim)
+        text_mask_expanded = text_mask.unsqueeze(-1).expand(-1, -1, embed_dim).to(inputs_embeds.device)
+        pad_mask_expanded = pad_mask.unsqueeze(-1).expand(-1, -1, embed_dim).to(inputs_embeds.device)
```
Comment (ArthurZucker) [inline src/transformers/models/paligemma/modeling_paligemma.py:305]: Ok this part I am not sure, the device that accelerate chooses for the input ids is usually automatically deduced? 
cc @SunMarc or @muellerzr ?

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 3]
```diff
@@ -301,14 +301,15 @@ def _merge_input_ids_with_image_features(
...
         # insert padding and text token embeddings
         final_embedding = torch.where(text_mask_expanded, inputs_embeds, final_embedding)
         final_embedding = torch.where(pad_mask_expanded, torch.zeros_like(final_embedding), final_embedding)
         # insert image embeddings - the image mask is always less or equal to the sentence in length
         final_embedding = final_embedding.masked_scatter(
-            image_mask.unsqueeze(-1).expand_as(final_embedding), scaled_image_features
+            image_mask.unsqueeze(-1).expand_as(final_embedding).to(device=final_embedding.device),
+            scaled_image_features.to(device=final_embedding.device, dtype=final_embedding.dtype),
         )
```
Comment (ArthurZucker) [inline src/transformers/models/paligemma/modeling_paligemma.py:313]: this makes sense tho, because you can split the encoder on another device so need to bring the image features to this device

File: src/transformers/models/paligemma/modeling_paligemma.py [hunk 4]
```diff
@@ -329,10 +330,12 @@ def _merge_input_ids_with_image_features(
...
                 mask_length = attention_mask.shape[-1]
-                padding_mask = causal_mask[:, :, :, :mask_length] + attention_mask[:, None, None, :]
+                padding_mask = causal_mask[:, :, :, :mask_length] + attention_mask[:, None, None, :].to(
+                    causal_mask.device
+                )
                 # unmask the prefill
                 causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
-                    token_type_ids[:, None, None, :] == 0, 0
+                    token_type_ids[:, None, None, :].to(causal_mask.device) == 0, 0
                 )
```
Comment (ArthurZucker) [inline src/transformers/models/paligemma/modeling_paligemma.py:339]: this one does not make sense to me

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 1 above)
Comment (ArthurZucker) [inline src/transformers/models/paligemma/modeling_paligemma.py:490]: this could make sense, but it's weird that all the other models don't need this. Not sure how the device of the labels is chosen in accelerate.

**Review by ArthurZucker — COMMENTED** [review]
thanks, will ping offline our accelerate experts I want to understand a bit better what's going on + why our tests did not catch this!

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 4 above)
Comment (molbap) [reply src/transformers/models/paligemma/modeling_paligemma.py:339]: with the masked_fill, you need both tensors to be on the same device, right?

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 1 above)
Comment (molbap) [reply src/transformers/models/paligemma/modeling_paligemma.py:490]: I think it's done like this in `Llava` as well for instance:
```python
shift_logits = logits[..., :-1, :][shift_attention_mask.to(logits.device) != 0].contiguous()
shift_labels = labels[..., 1:][shift_attention_mask.to(labels.device) != 0].contiguous()
```

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 3 above)
Comment (molbap) [reply src/transformers/models/paligemma/modeling_paligemma.py:313]: yes, that was the main one that was causing issues

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/models/paligemma/modeling_paligemma.py:339]: Ah, sorry, I read to fast. 
So token type ids's device is not correctly inferred ? Or are we not creating the causal mask on the correct device? It should be created on the input or attention mask's device for consistency, since when it's used accelerate will transfer it accordingly I think.

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/paligemma/modeling_paligemma.py:490]: ok!

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 4 above)
Comment (molbap) [reply src/transformers/models/paligemma/modeling_paligemma.py:339]: hmm, I think we are setting the causal_mask to the correct device. It's the token_type_id device that is indeed not correctly inferred

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 2 above)
Comment (SunMarc) [reply src/transformers/models/paligemma/modeling_paligemma.py:305]: Yeah, when `input_ids` is called in the forward of `PaliGemmaForConditionalGeneration`, it will be moved the [main device](https://github.com/huggingface/accelerate/blob/0553483638fcde878158acef9cfd6dbb62757515/src/accelerate/big_modeling.py#L391) thanks to a pre_forward hook (e.g. cuda:0) . However, since `inputs_embeds` is not necessarily on the same device as `input_ids` (e.g. cuda:1). Hence we can't really make sure that they are both on the same device.

Comment (molbap) [reply src/transformers/models/paligemma/modeling_paligemma.py:305]: Right, that's what happened when I tried on a multi-gpu setting with `device_map` to `auto`, they were on different devices. And this hook can't apply to `inputs_embeds` because it is created later, is that correct?

Comment (SunMarc) [reply src/transformers/models/paligemma/modeling_paligemma.py:305]: Yeah it's right !

File: src/transformers/models/paligemma/modeling_paligemma.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/models/paligemma/modeling_paligemma.py:339]: But from the comment of @SunMarc I would suspect both devices to be the same no

Comment (molbap) [reply src/transformers/models/paligemma/modeling_paligemma.py:339]: hmm, doesn't seem like it, tried on a multi-gpu env with device_map to auto and got 
image

Comment (molbap) [reply src/transformers/models/paligemma/modeling_paligemma.py:339]: @SunMarc if you have an idea here  - `token_type_ids` is created by the processor along with `input_ids` and passed to the `forward` normally

Comment (SunMarc) [reply src/transformers/models/paligemma/modeling_paligemma.py:339]: So, from the code, I see that `token_type_ids` is indeed on the same device as `input_ids`. However, since you created the `causal_mask` to be on the same device as `inputs_embeds.device`, `token_type_ids` and `input_ids` might not be on the same device. 
```python 
            causal_mask = torch.full(
                (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
            )
```
where
` dtype, device = inputs_embeds.dtype, inputs_embeds.device`

Comment (molbap) [reply src/transformers/models/paligemma/modeling_paligemma.py:339]: alright, thanks! in that case, we can keep it as it is? The other way is to create the causal mask on the `input_ids.device`, I'm not sure if one is better than the other - `inputs_embeds` is much larger in general

**Review by ArthurZucker — APPROVED** [review]

**Comment by grahamannett** [comment]
fwiw most of the lines in here are nearly identical to the changes i have done locally as well besides the `final_embedding` related one which i believe can be done with only 1 cast but didnt think too deeply about it

**Comment by molbap** [comment]
@grahamannett , good to know. For `final_embedding` it's also to fix the bf16 dtype mismatch.
